cert/spec.go

```diff
@@ -140,48 +127,59 @@ func (spec *Spec) identity() (*core.Identity, error) {
 	return ident, nil
 }
 
-func newSpecFromPath(path string) (*Spec, error) {
+// loadFromPath load and fill this spec from the given pathway
+// If this invocation returns an error, the spec instance should be discarded
+// and recreated.
+func (spec *Spec) loadFromPath(path string) error {
 	in, err := ioutil.ReadFile(path)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	specStat, err := os.Stat(path)
 	if err != nil {
 		// Hit the race; we read the file but someone wiped it.
-		return nil, err
-	}
-	var spec = &Spec{
-		Request:  csr.New(),
-		Path:     path,
-		loadTime: specStat.ModTime(),
+		return err
 	}
 
 	switch filepath.Ext(path) {
 	case ".json":
-		err = json.Unmarshal(in, &spec)
+		err = strictJSONUnmarshal(in, &spec)
 	case ".yml", ".yaml":
 		err = yaml.UnmarshalStrict(in, &spec)
 	default:
 		err = fmt.Errorf("cert: unrecognised spec file format for %s", path)
 	}
 
-	return spec, err
+	if err == nil {
+		spec.loadTime = specStat.ModTime()
+		spec.Path = path
+	}
+	return err
 }
 
 // Load reads a spec from a JSON configuration file.
 func Load(path, remote string, before time.Duration, defaultServiceManager string, strict bool) (*Spec, error) {
-	spec, err := newSpecFromPath(path)
+	var spec = &Spec{
+		Request:            csr.New(),
+		ServiceManagerName: defaultServiceManager,
+	}
+	spec.CA.Remote = remote
+	err := spec.loadFromPath(path)
 	if err != nil {
 		return nil, err
 	}
 
 	if spec.CA.Remote == "" {
-		spec.CA.Remote = remote
+		return nil, errors.New("no remote specified in authority (either in the spec or in the certmgr config)")
 	}
 
-	if spec.CA.Remote == "" {
-		return nil, errors.New("cert: no remote specified in authority (either in the spec or in the certmgr config)")
+	if (spec.Cert == nil) != (spec.Key == nil) {
+		return nil, errors.New("if either cert or key are defined, both fields must be defined as must request to successfully write the keypair to disk")
```

Comment:
`s/defined as must request to/defined, as must request, to` to disambiguate "cert and key being defined as 'must request'" from "cert and key must be defined _in addition to_ request".